### PR TITLE
HPCC-18950 Roxie aggregate performance-suite issues

### PR DESCRIPTION
**Pull request by richardkchapman**
Move allocation of rowbuilder objects onto the Roxie heap
(which is really more appropriate anyway) to avoid heap
fragmentation issues using glibc malloc.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Perf suite / obt / smoketest.

Timing the particular test that triggered this investigation showed a significant impact (20% or more)

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith @ghalliday Please review

File: common/thorhelper/thorcommon.ipp [hunk 1]
```diff
@@ -119,6 +119,29 @@ private:
...
+public:
+    IMPLEMENT_IINTERFACE
+
+    virtual size32_t getRecordSize(const void *rec)         { return sizeof(T); }
+    virtual size32_t getMinRecordSize() const               { return sizeof(T); }
+    virtual size32_t getFixedSize() const                   { return sizeof(T); }
+    virtual void toXML(const byte * self, IXmlWriter & out) { }
+    virtual unsigned getVersion() const                     { return OUTPUTMETADATA_VERSION; }
+    virtual unsigned getMetaFlags()                         { return 0; }
+    virtual void destruct(byte * self)  {}
```
Comment (ghalliday) [inline common/thorhelper/thorcommon.ipp]: Maybe you should implement this as T::~T(self);

File: common/thorhelper/thorcommon.ipp [hunk 2]
```diff
@@ -119,6 +119,29 @@ private:
     unsigned metaFlags;
 };
 
+template<class T> class CFixedMeta : implements IOutputMetaData, public CInterface
```
Comment (ghalliday) [inline common/thorhelper/thorcommon.ipp]: IInterfaceOf<IOutputMetaData> is probably slightly better

**Comment by jakesmith** [comment]
05bc_summaryallhash was the query that was pathologically slow in Thor [sometimes] before it's approach to aggregating was refactored (See HPCC-17181).
It used to use RowAggregate and therefore create a AggregateRowBuilder per aggregate row, which particularly killed it when reading them back and merging.

Wonder if worth moving/commoning up on Thor's implementation ? (see class CAggregateHT in thhashdistributeslave.cpp)

**Comment by richardkchapman** [comment]
@jakesmith  It probably IS worth commoning up - but I may leave that for a separate Jira. This one is worth merging as-is I suspect, not least because it has some useful examples/precedent for use of roxiemem in place of new ...

**Pull request merged**